compiler/src/iree/compiler/Codegen/Common/ExtractAddressComputation.h

```diff
@@ -0,0 +1,86 @@
+// Copyright 2023 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+#ifndef IREE_COMPILER_CODEGEN_COMMON_EXTRACTADDRESSCOMPUTATION_H_
+#define IREE_COMPILER_CODEGEN_COMMON_EXTRACTADDRESSCOMPUTATION_H_
+
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/IR/PatternMatch.h"
+
+using namespace mlir;
+
+namespace mlir {
+namespace iree_compiler {
+
+/// Rewrite a load-like op so that all its indices are zeros.
+/// E.g., %ld = memref.load %base[%off0]...[%offN]
+/// =>
+/// %new_base = subview %base[%off0,.., %offN][1,..,1][1,..,1]
+/// %ld = memref.load %new_base[0,..,0] :
+///    memref<1x..x1xTy, strided<[1,..,1], offset: ?>>
+///
+/// `getSrcMemRef` returns the source memref for the given load-like operation.
+///
+/// Using the given rewriter, `rebuildOpFromAddressAndIndices` creates a new
+/// LoadLikeOp that reads from srcMemRef[indices].
+/// The returned operation will be used to replace loadOp.
+template <typename LoadLikeOp, Value (*getSrcMemRef)(LoadLikeOp),
```

Comment:
Would it be possible to not make this templated and move the implementation out of a header file? Fine if this can be done later as well. Seems like this is calling for a `LoadOpInterface`. So not a blocking requirement.

Reply:
Yes, the plan is to add an interface and get rid of this.
The interface itself will like spark some conversation in LLVM upstream and I didn't want to wait for that to happen.